### ISSUE TITLE
listviews should remind their state

### ISSUE THREAD

**Issue by georgtroska** [Martin-Gleiss/smartvisu#921]
Hi,
thank you very much for this great project.

There is one thing that makes me really agressive :-) After a reload the listviews do not stay in their last state.

In particular the attribute data-collapsed can be set individually in the html, but its stored nowhere. Maybe I'm doing something wrong, but in my case the rooms_menu goes back into its initial state after opening a collaped section and opening a page in there. This means the opened section of the listview is collased again. 

Maybe you can help me to fix. Not sure if it is a misconfiguration on my side or if a state-remind is not taken into account so far
Thanks
Georg

**Comment by wvhn** [comment]
As the rooms_menu is included in every new page as a module it will not remember it‘s state from a previous page.
You can however control the collapsed sections programmatically by linking them to the pages contained. View index.html in the docu pages to see how this is achieved.

Example: if there is a collapsible section with rooms called „living“, „kitchen“ and „dining“ you can open this section every time one of these pages is being displayed by checking the id of the displayed page:

```
<div data-role="collapsible"{{ page in ['living','kitchen’, dining'] ? ' data-collapsed="false"' }} …>
```
Regards
Wolfram

**Comment by wvhn** [comment]
@georgtroska 
any comments on this? Did this solve the problem?
I'd like to close this issue and - if necessary - continue the discussion in the forum as this may be of interest to many users.

Regards 
Wolfram

**Comment by georgtroska** [comment]
Sorry – other things to do in the moment – please give me some more days

 

Georg

 

Von: wvhn ***@***.***> 
Gesendet: Donnerstag, 21. September 2023 13:29
An: Martin-Gleiss/smartvisu ***@***.***>
Cc: georgtroska ***@***.***>; Mention ***@***.***>
Betreff: Re: [Martin-Gleiss/smartvisu] listviews should remind their state (Issue #921)

 

@georgtroska <https://github.com/georgtroska> 
any comments on this? Did this solve the problem?
I'd like to close this issue and - if necessary - continue the discussion in the forum as this may be of interest to many users.

Regards
Wolfram

**Comment by georgtroska** [comment]
@wvhn
Hi, I just tried it out - thank you very much for your hint.
It works much much better as anything before :-)

Unfortunately, it seems not to be running 100% smoothly, let me give you a brief inside of my code:
`
div data-role="collapsible-set" style="margin-top: -15px;" data-inset="false" data-mini="true" data-content-theme="a">
	<div data-role="collapsible"{{page in ['room_Aussenbereich','room_Zentral'] ? ' data-collapsed="false"' }} data-theme="c" data-content-theme="a">
		<h3>Haus</h3>
		<ul data-role="listview">
			<li data-icon="false">
				<a href="index.php?page=room.Zentral">
					<img class="icon" src="{{ icon0 }}control_all_on_off.svg" />

					<h3>Zentralfunktionen</h3>
					
				</a>
			</li>
			<li data-icon="false">
				<a href="index.php?page=room.Aussenbereich">
					<img class="icon" src="{{ icon0 }}control_building_outside.svg" />

					<h3>Außenbereich</h3>
					<div class="ui-li-aside"></div>
				</a>
			</li>
		</ul>
	</div>
	<div data-role="collapsible"{{page in ['room_EG_Flur','room_WC','room_Kueche','room_Esszimmer','room_Kino','room_Wohnzimmer'] ? ' data-collapsed="false"' }} data-theme="c" data-content-theme="a">
		<h3>EG</h3>
---some more---
	</div>
</div>
`

When I'm clicking through all my pages it runs, but every second time or so 'Haus.Aussenbereich' is shown in the wrong state.... actually it will be shown in the state which was active before

Any hint?

Thanks
Georg

**Comment by wvhn** [comment]
@georgtroska 
the twig code is executed on server side before the html page is rendered in the browser. So it leaves just a 'data-collapsed="false"' in the static part of the page.  After html rendering the dynamic page manipulation is done by jQuery mobile. jQm controls the collapsibles and remembers their states. For performance reasons we use jQm multipage mode (AJAX navigation) which keeps all visited pages in the memory with their current states. 

So if you browse through your room menu and then go back to a page on which you had toggled the listview before, you will see the last active state instead of the state defined in twig. I think that is what you are experiencing right now.

So to solve this we need to reset the listviews to their original states by jQm code. I have made a quick and dirty approach in the pagecontainershow event handler.  If you want to try this insert the following code into ./pages/base/root.html before line 239 (master v3.3.1) or line 240 (develop v3.3.b):
```
$(sv.activePage).find('.secondary [data-collapsed="false"] .ui-collapsible-heading-collapsed').parents().first().collapsible('expand');
```
If this works for you I can integrate this as a config option.

Regards 
Wolfram

**Comment by wvhn** [comment]
@georgtroska 
Have you been able to test the patch and does it work as expected?

Regards
Wolfram

**Comment by wvhn** [comment]
No answer - therefore closed with #927
